bionic/code_hasher.py

```diff
@@ -0,0 +1,201 @@
+            if isinstance(obj, list):
+                type_prefix = TypePrefix.LIST
+            else:
+                type_prefix = TypePrefix.TUPLE
+            byte_obj = str(len(obj)).encode()
+            self._write_to_hash_object(
+                type_prefix=type_prefix, byte_obj=byte_obj,
+            )
+            for elem in obj:
+                self._update_with_checks(elem)
+
+        elif isinstance(obj, dict):
+            self._write_to_hash_object(
+                type_prefix=TypePrefix.DICT, byte_obj=str(len(obj)).encode(),
+            )
+            for key, elem in obj.items():
+                self._update_with_checks(key)
+                self._update_with_checks(elem)
+
+        elif inspect.isroutine(obj):
+            self._write_to_hash_object(type_prefix=TypePrefix.ROUTINE)
+            if obj.__defaults__ is not None:
+                for d in obj.__defaults__:
+                    self._update_with_checks(d)
+
+            self._hash_code(obj.__code__)
+
+        elif inspect.iscode(obj):
+            self._write_to_hash_object(type_prefix=TypePrefix.CODE)
+            self._hash_code(obj)
+
+        else:
+            # TODO: Verify that we hash all Python constant types.
+            self._write_to_hash_object(type_prefix=TypePrefix.DEFAULT)
+            message = oneline(
+                f"""
+                Found a constant {obj!r} of type {type(obj)!r} that
+                Bionic doesn't know how to hash. This is most likely a
+                bug in Bionic. Please raise a new issue at
+                https://github.com/square/bionic/issues to let us know.
+                """
+            )
+            warnings.warn(message)
+
+    def _hash_code(self, code):
+        # TODO: Find references for the code and analyze references.
+        self._update_with_checks(code.co_code)
+
+        # Hash constants used in the code object, but ignore names of lambdas.
+        consts = [
+            const
+            for const in code.co_consts
+            if not isinstance(const, str) or not const.endswith(".<lambda>")
+        ]
+        self._update_with_checks(consts)
+
+
+class TypePrefix(Enum):
+    """
+    Represents a unique value for each type that CodeHasher hashes that
+    is prefixed to avoid collision between same encoded values.
+
+    If you change the prefix of any type, or add a new prefix, it can
+    change the encoding of objects. Since Bionic uses these encodings
+    to detect changes, any changes to encoding will make caching
+    backwards-incompatible. Update CACHE_SCHEMA_VERSION to update cache
+    scheme if you change the encoding.
+    """
+
+    BYTES = b"AA"
+    BYTEARRAY = b"AB"
+    NONE = b"AC"
+    STRING = b"AD"
+    INT = b"AE"
+    FLOAT = b"AF"
+    BOOL = b"AG"
+    LIST = b"AH"
+    TUPLE = b"AI"
+    DICT = b"AJ"
+    ROUTINE = b"AK"
+    CODE = b"AL"
+    CIRCULAR_REF = b"AM"
+    DEFAULT = b"ZZ"
+
+
+def get_size_in_bytes(byte_obj):
```

Comment:
The name `get_size_in_bytes` makes me think this will return an integer indicating the size of an object, measured in bytes. `get_size_as_bytes` might be clearer?